### PR TITLE
Add quantum state discrimination tutorial

### PR DESCRIPTION
**Pull request by odow**
x-ref https://github.com/jump-dev/JuMP.jl/issues/2348#issuecomment-1446206831

Also fixes https://github.com/jump-dev/JuMP.jl/pull/3245#issuecomment-1447048453

Still a WIP, because I don't really know enough physics to write about the problem we're solving, or what the solution represents.

Preview: https://jump.dev/JuMP.jl/previews/PR3250/tutorials/conic/quantum_discrimination/

### CODE REVIEW

File: docs/src/tutorials/conic/quantum_discrimination.jl [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+# A `d`-dimensional quantum state, ``\rho``, can be defined by a complex-valued
+# Hermitian matrix with a trace of `1`. Assume we have `N` `d`-dimensional
+# quantum states, ``\{\rho_i}_{i=1}^n``, each of which is equally likely.
+
+# The goal of the Quantum state discrimination problem is to choose a set of
+# positive-operator-valued-measures (POVMs), ``E_i`` such that if we observe
+# ``E_i`` then the most probable state that we are in is ``\rho_i``.
+
+# Each POVM ``E_i`` is a complex-valued Hermitian matrix, and there is a
+# requirement that ``\sum\limits E_i = \mathbf{I}``.
```
Comment (odow) [inline docs/src/tutorials/conic/quantum_discrimination.jl:33]: This is what I've inferred by reading https://arxiv.org/pdf/1707.02571.pdf.

But I don't actually understand what a POVM is in practice...

Comment (jd-foster) [reply docs/src/tutorials/conic/quantum_discrimination.jl]: And neither does vale by the look of it :)

File: docs/src/tutorials/conic/quantum_discrimination.jl [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+optimize!(model)
+solution_summary(model)
+
+# The POVMs are:
+
+solution = [value.(e) for e in E]
+
+# ## Alternative formulation
+
+# The formulation above includes `n` Hermitian matrices, and a set of linear
```
Comment (jd-foster) [inline docs/src/tutorials/conic/quantum_discrimination.jl:109]: Did you mean to switch to lowercase `n` from here on?

File: docs/src/tutorials/conic/quantum_discrimination.jl [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import LinearAlgebra
+import SCS
+
+# ## Formulation
+
+# A `d`-dimensional quantum state, ``\rho``, can be defined by a complex-valued
+# Hermitian matrix with a trace of `1`. Assume we have `N` `d`-dimensional
+# quantum states, ``\{\rho_i\}_{i=1}^n``, each of which is equally likely.
+
+# The goal of the Quantum state discrimination problem is to choose a set of
```
Comment (jd-foster) [inline docs/src/tutorials/conic/quantum_discrimination.jl:28]:
```suggestion
# The goal of the quantum state discrimination problem is to choose a set of
```

File: docs/src/tutorials/conic/quantum_discrimination.jl (hunk 2 above)
Comment (odow) [reply docs/src/tutorials/conic/quantum_discrimination.jl]: Fixed

File: docs/src/tutorials/conic/quantum_discrimination.jl (hunk 1 above)
Comment (odow) [reply docs/src/tutorials/conic/quantum_discrimination.jl]: Fixed vale

File: docs/src/tutorials/conic/quantum_discrimination.jl [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+# quantum states, ``\{\rho_i\}_{i=1}^N``, each of which is equally likely.
+
+# The goal of the Quantum state discrimination problem is to choose a set of
+# positive-operator-valued-measures (POVMs), ``E_i`` such that if we observe
+# ``E_i`` then the most probable state that we are in is ``\rho_i``.
+
+# Each POVM ``E_i`` is a complex-valued Hermitian matrix, and there is a
+# requirement that ``\sum\limits_{i=1}^N E_i = \mathbf{I}``.
+
+# To choose the set of POVMs, we want to maximize the probability that we guess
```
Comment (jd-foster) [inline docs/src/tutorials/conic/quantum_discrimination.jl:35]:
```suggestion
# When choosing a POVM, we want to maximize the probability that we guess
```

File: docs/src/tutorials/conic/quantum_discrimination.jl [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    Max,
+    sum(real(LinearAlgebra.tr(ρ[i] * E[i])) for i in 1:N) / N,
+)
+
+# Now we optimize:
+
+optimize!(model)
+solution_summary(model)
+
+# The POVMs are:
```
Comment (jd-foster) [inline docs/src/tutorials/conic/quantum_discrimination.jl:103]:
```suggestion
# The optimal POVM is:
```

File: docs/styles/Vocab/JuMP-Vocab/accept.txt [hunk 6]
```diff
@@ -78,6 +78,7 @@ nl|NL
 overfitting
 parameteriz(ing|ation)
 perp
+POVMs
```
Comment (jd-foster) [inline docs/styles/Vocab/JuMP-Vocab/accept.txt:81]:
```suggestion
POVM
```

**Review by jd-foster — COMMENTED** [review]
A POVM is the set of PSD matrices over the states 1:N, not the individual matrices.

File: docs/src/tutorials/conic/quantum_discrimination.jl [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+# Each POVM ``E_i`` is a complex-valued Hermitian matrix, and there is a
+# requirement that ``\sum\limits_{i=1}^N E_i = \mathbf{I}``.
+
+# To choose the set of POVMs, we want to maximize the probability that we guess
+# the quantum state corrrectly. This can be formulated as the following
+# optimization problem:
+
+# ```math
+# \begin{aligned}
+# \max\limits_{E} \;\; & \mathbb{E}_i[ tr(\rho_i \times E_i)] \\
```
Comment (jd-foster) [inline docs/src/tutorials/conic/quantum_discrimination.jl:41]:
```suggestion
# \max\limits_{E} \;\; & \mathbb{E}_i[ \operatorname{tr}(\rho_i \times E_i)] \\
```

File: docs/src/tutorials/conic/quantum_discrimination.jl [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+# \text{s.t.}     \;\; & \sum\limits_{i=1}^N E_i = \mathbf{I} \\
+#                      & E_i \succeq 0 \forall i = 1,\ldots,N.
```
Comment (jd-foster) [inline docs/src/tutorials/conic/quantum_discrimination.jl:43]:
```suggestion
#                      & E_i \succeq 0 \; \forall i = 1,\ldots,N.
```

File: docs/src/tutorials/conic/quantum_discrimination.jl [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+# This Source Code Form is subject to the terms of the Mozilla Public License   #src
+# v.2.0. If a copy of the MPL was not distributed with this file, You can       #src
+# obtain one at https://mozilla.org/MPL/2.0/.                                   #src
+
+# # Quantum state discrimination
+
+# This tutorial solves the problem of [quantum state discrimination](https://en.wikipedia.org/wiki/Quantum_state_discrimination).
+
+# The purpose of this tutorial to demonstrate how to solve problems involving
+# complex-valued decision variables and the [`HermitianPSDCone`(@ref). See
```
Comment (metab0t) [inline docs/src/tutorials/conic/quantum_discrimination.jl:11]: [`HermitianPSDCone`](@ref)

Comment (odow) [inline docs/src/tutorials/conic/quantum_discrimination.jl:11]:
```suggestion
# complex-valued decision variables and the [`HermitianPSDCone`](@ref). See
```

**Comment by jd-foster** [comment]
This all looks great to me. My only last question is whether you wanted to seed the random generation?

**Comment by odow** [comment]
> My only last question is whether you wanted to seed the random generation?

Nah. It'll be okay.

**Comment by araujoms** [comment]
The task is the following: you are given an unknown quantum state, promised to come from a known set of n states (sampled with uniform probability, for simplicity). You need to find out which quantum state it is. The objective of the SDP is the probability of success, \sum_i p( guess i| state i)/n. The matrices E[i] are the POVM elements, encoding the optimal measurement you can do.

**Comment by araujoms** [comment]
More specific comments: one is interested in the value of the objective, not on the POVM elements, so I'd print the objective instead. Also, in the problem formulation I'd write the objective as \sum_i \operatorname{tr}(\rho_i  E_i)/N instead of \mathbb{E}_i[ \operatorname{tr}(\rho_i  E_i)].

Also worth remarking is that for N=2 the problem is trivial, with solution given by 0.5 + 0.25*sum(svdvals(rho[1]-rho[2])).

**Comment by odow** [comment]
Updated with @araujoms suggestions. I don't know if we need to go into detail about the theory behind the model. People interested in `HermitianPSDCone` will probably already have the necessary background.

**Comment by odow** [comment]
Merging since we seem to have converged on this. Any other changes can happen in a new PR.

**Pull request merged**